driver-async/src/test/functional/com/mongodb/async/client/RetryableWritesProseTest.java

```diff
@@ -0,0 +1,224 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.mongodb.async.client;
+
+import com.mongodb.Block;
+import com.mongodb.MongoException;
+import com.mongodb.MongoNotPrimaryException;
+import com.mongodb.MongoServerException;
+
+import com.mongodb.async.FutureResultCallback;
+import com.mongodb.async.SingleResultCallback;
+import com.mongodb.connection.ServerSettings;
+import com.mongodb.internal.connection.TestCommandListener;
+import org.bson.BsonDocument;
+import org.bson.BsonInt32;
+import org.bson.Document;
+import org.junit.After;
+import org.junit.AfterClass;
+import org.junit.Before;
+import org.junit.BeforeClass;
+import org.junit.Test;
+
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+
+import static com.mongodb.ClusterFixture.isDiscoverableReplicaSet;
+import static com.mongodb.ClusterFixture.serverVersionAtLeast;
+import static org.junit.Assert.fail;
+import static org.junit.Assume.assumeTrue;
+
+import static com.mongodb.ClusterFixture.getDefaultDatabaseName;
+import static com.mongodb.async.client.Fixture.getMongoClientBuilderFromConnectionString;
+
+// See https://github.com/mongodb/specifications/tree/master/source/retryable-writes/tests/README.rst#replica-set-failover-test
+public class RetryableWritesProseTest extends DatabaseTestCase {
+    private MongoClient clientUnderTest;
+    private MongoClient failPointClient;
+    private MongoClient stepDownClient;
+    private final TestCommandListener commandListener;
+    private final String databaseName;
+    private final String collectionName;
+    private final String description;
+
+    public RetryableWritesProseTest() {
+        this.description = "Replica set failover test for retryable writes";
+        this.databaseName = getDefaultDatabaseName();
+        this.collectionName = "test";
+        this.commandListener = new TestCommandListener();
+    }
+
+    @BeforeClass
+    public static void beforeClass() {
+    }
+
+    @AfterClass
+    public static void afterClass() {
+    }
+
+    @Before
+    @Override
+    public void setUp() {
+        assumeTrue(canRunTests());
+
+        setUpClientUnderTest();
+        setUpFailPointClient();
+        setUpStepDownClient();
+    }
+
+    @After
+    public void cleanUp() {
+        if (failPointClient != null) {
+            MongoDatabase failPointAdminDB = failPointClient.getDatabase("admin");
+            FutureResultCallback<Document> futureResultCallback = new FutureResultCallback<Document>();
+            failPointAdminDB.runCommand(
+                    new Document("configureFailPoint", "stepdownHangBeforePerformingPostMemberStateUpdateActions")
+                            .append("mode", "off"), futureResultCallback);
```

Comment:
With the async driver, all methods that take a callback run asynchronously and the results are passed to the callback.
To make this code work as expected call `futureResultCallback.get` after each callback.  That will block and wait for the result or throws an error if the operation failed.